kernel/src/grant.rs

```diff
@@ -240,75 +449,202 @@ impl<'a> GrantUpcallTable<'a> {
+struct SavedUpcall {
+    appdata: usize,
+    fn_ptr: Option<NonNull<()>>,
 }
 
-/// Subscribe to an upcall by saving the upcall in the grant region for the
-/// process and returning the existing upcall for the same UpcallId.
-pub(crate) fn subscribe(
-    process: &dyn Process,
-    upcall: Upcall,
-) -> Result<Upcall, (Upcall, ErrorCode)> {
-    let grant_num = match process.lookup_grant_from_driver_num(upcall.upcall_id.driver_num) {
-        Ok(grant_num) => grant_num,
-        Err(e) => return Err((upcall, e.into())),
-    };
+impl Default for SavedUpcall {
+    fn default() -> Self {
+        Self {
+            appdata: 0,
+            fn_ptr: None,
+        }
+    }
+}
 
-    // Check if the grant has been allocated, and if not we cannot handle the
-    // subscribe call.
-    if let Some(is_allocated) = process.grant_is_allocated(grant_num) {
-        if !is_allocated {
-            return Err((upcall, ErrorCode::NOMEM));
+/// A minimal representation of a read-only allow from app, used for
+/// storing a read-only allow in a process' kernel managed grant space
+/// without wasting memory duplicating information such as process ID.
+#[repr(C)]
+struct SavedAllowRo {
+    ptr: *const u8,
+    len: usize,
+}
+
+impl Default for SavedAllowRo {
+    fn default() -> Self {
+        Self {
+            ptr: core::ptr::null(),
+            len: 0,
         }
-    } else {
-        // Process is no longer active, this case will never happen.
-        return Err((upcall, ErrorCode::FAIL));
     }
+}
 
-    // Return early if no grant.
-    let grant_ptr = match process.enter_grant(grant_num) {
-        Ok(grant_ptr) => grant_ptr,
-        Err(_) => return Err((upcall, ErrorCode::NOMEM)),
+/// A minimal representation of a read-write allow from app, used for
+/// storing a read-write allow in a process' kernel managed grant space
+/// without wasting memory duplicating information such as process ID.
+#[repr(C)]
+struct SavedAllowRw {
+    ptr: *mut u8,
+    len: usize,
+}
+
+impl Default for SavedAllowRw {
+    fn default() -> Self {
+        Self {
+            ptr: core::ptr::null_mut(),
+            len: 0,
+        }
+    }
+}
+
+/// Write the default value of T to every element of the array.
+///
+/// # Safety
+///
+/// The pointer must be well aligned and point to allocated memory that is writable
+/// for `size_of::<T> * num` bytes. The memory does not need to be initialized yet. If
+/// it already does contain initialized memory, then those contents will be overwritten
+/// without being `Drop`ed first.
+unsafe fn write_default_array<T: Default>(base: *mut T, num: usize) {
+    for i in 0..num {
+        base.add(i).write(T::default());
+    }
+}
+
+/// Lifetime of guard represents the lifetime a grant is help "open". On Drop, we leave grant
+struct GrantEnterLifetimeGuard<'a>(&'a dyn Process, usize);
+
+impl Drop for GrantEnterLifetimeGuard<'_> {
+    fn drop(&mut self) {
+        self.0.leave_grant(self.1);
+    }
+}
```

Comment:
The main concern is that safe code (which includes untrusted capsule code) can call `core::mem::forget` and get rid of the `GrantEnterLifetimeGuard` without calling `Drop::drop`. In this case, that would cause `self.0.leave_grant` to not be called. Based on a quick look (I'm still not very familiar with the kernel crate), it doesn't look like that would lead to memory unsafey, so I think it is fine.

I'm not entirely convinced that Tock does not need to worry about unwinding, though. I don't think implementing unwinding is terribly difficult, plus if we do some sort of host-based testing (e.g. running Tock kernel code under Miri on a Linux system) then unwinding will be possible.

Reply:
Agreed that it is undesirable to `mem::forget` this object, but it is not unsound.

Marking as resolved and we can talk about unwinding concerns in a future PR/issue since there aren't any immediate concerns with this code delta.